pkg/clustersserver/clustersserver.go

```diff
@@ -2,21 +2,184 @@ package clustersserver
 
 import (
 	"context"
+	"errors"
+	"fmt"
 
+	"k8s.io/client-go/tools/clientcmd"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta1"
+	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1beta1"
+	notificationv1 "github.com/fluxcd/notification-controller/api/v1beta1"
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
 	pb "github.com/fluxcd/webui/pkg/rpc/clusters"
-	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/apimachinery/pkg/runtime"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 )
 
 type Server struct {
-	ClientConfig *clientcmdapi.Config
+	LoadingRules *clientcmd.ClientConfigLoadingRules
+}
+
+func (s *Server) newKubeClient(kubeContext string) (client.Client, error) {
+	configOverrides := clientcmd.ConfigOverrides{CurrentContext: kubeContext}
+
+	restCfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		s.LoadingRules,
+		&configOverrides,
+	).ClientConfig()
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create rest config: %w", err)
+	}
+
+	scheme := apiruntime.NewScheme()
+	_ = appsv1.AddToScheme(scheme)
+	_ = corev1.AddToScheme(scheme)
+	_ = rbacv1.AddToScheme(scheme)
+	_ = sourcev1.AddToScheme(scheme)
+	_ = kustomizev1.AddToScheme(scheme)
+	_ = helmv2.AddToScheme(scheme)
+	_ = notificationv1.AddToScheme(scheme)
+
+	kubeClient, err := client.New(restCfg, client.Options{
+		Scheme: scheme,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("kubernetes client initialization failed: %w", err)
+	}
+
+	return kubeClient, nil
 }
 
 func (s *Server) ListContexts(ctx context.Context, msg *pb.ListContextsReq) (*pb.ListContextsRes, error) {
+	cfg, err := s.LoadingRules.Load()
+
+	if err != nil {
+		return nil, fmt.Errorf("could not load rules %w", err)
+	}
 
 	ctxs := []*pb.Context{}
-	for _, c := range s.ClientConfig.Contexts {
+	for _, c := range cfg.Contexts {
 		ctxs = append(ctxs, &pb.Context{Name: c.Cluster})
 	}
 
-	return &pb.ListContextsRes{Contexts: ctxs}, nil
+	return &pb.ListContextsRes{Contexts: ctxs, CurrentContext: cfg.CurrentContext}, nil
+}
+
+func (s *Server) ListKustomizations(ctx context.Context, msg *pb.ListKustomizationsReq) (*pb.ListKustomizationsRes, error) {
+	client, err := s.newKubeClient(msg.ContextName)
```

Comment:
I think we need to use a global client to take advantage of the builtin cache.

Reply:
It wasn't clear how to change the context of an existing client, since each request from the UI can be for a different context. I will look at modifying a client that is already instantiated.